File: chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster #静态坐标发布器
from geometry_msgs.msg import TransformStamped #消息接口
from tf_transformations import quaternion_from_euler #欧拉角转四元数
import math #角度转弧度函数

class StaticTFBroadcaster(Node):
    def __init__(self):
        super().__init__('static_tf_broadcaster') #初始化节点
        self.static_broadcaster_ = StaticTransformBroadcaster(self) #创建静态坐标发布器对象
        self.publish_static_tf() #调用发布静态坐标变换函数

        # self.timer = self.create_timer(1.0, self.broadcast_static_transform) #创建定时器，每秒调用一次广播函数

    def publish_static_tf(self):
        """
        发布静态坐标变换，从base_link到camera_link之间的坐标变换
        """
        transform = TransformStamped() #创建坐标变换消息对象
        transform.header.stamp = self.get_clock().now().to_msg() #获取当前时间戳
        transform.header.frame_id = 'base_link' #父坐标系
        transform.child_frame_id = 'camera_link' #子坐标系

        # 设置平移参数
        transform.transform.translation.x = 0.5 #x方向平移0.5
        transform.transform.translation.y = 0.3 #y方向平移0.3
        transform.transform.translation.z = 0.6 #z方向平移0.6

        # 设置旋转参数，使用欧拉角转四元数
        q = quaternion_from_euler(math.radians(180), 0, 0) #将角度转为弧度
        #旋转部分进行赋值
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]

        # 发布静态坐标变换
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f'发布静态TF:{transform}')


def main():
    rclpy.init() #初始化rclpy
    node = StaticTFBroadcaster() #创建节点对象
    rclpy.spin(node) #循环等待回调函数
    rclpy.shutdown() #关闭rclpy

File: chapt5/chapt5_ws/src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/static_transform_broadcaster.h" //静态坐标变换广播器
#include "geometry_msgs/msg/transform_stamped.hpp" //坐标变换消息
#include "tf2/LinearMath/Quaternion.h" //提供tf2::Quaternion类
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp" //消息类型转换函数

class StaticTFBroadcaster : public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_; //静态坐标变换广播器

public:
    StaticTFBroadcaster() : Node("static_tf_broadcaster")
    {
        // 创建静态坐标变换广播器
        broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_tf(); //发布静态坐标变换
        RCLCPP_INFO(this->get_logger(), "发布map坐标系到target_point坐标系的静态坐标变换");
    }

    void publish_tf()
    {
        geometry_msgs::msg::TransformStamped transform;

        transform.header.stamp = this->get_clock()->now(); //设置时间戳

        // 设置父坐标系和子坐标系
        transform.header.frame_id = "map"; //父坐标系
        transform.child_frame_id = "target_point"; //子坐标系

        // 设置平移参数
        transform.transform.translation.x = 5.0; //x方向平移5
        transform.transform.translation.y = 3.0; //y方向平移3
        transform.transform.translation.z = 0.0; //z方向平移0

        // 设置旋转参数，使用欧拉角转四元数
        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, 60*M_PI/180.0); //设置绕Z轴旋转60度
        transform.transform.rotation = tf2::toMsg(q);//将tf2::Quaternion类型转换为geometry_msgs::msg::Quaternion类型

        // 发布静态坐标变换
        broadcaster_->sendTransform(transform);
    }

};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}